src/components/ModalPayment/ModalMegaGam/ModalMegaGam.tsx

```diff
@@ -0,0 +1,373 @@
+  const { t } = useTranslation();
+  const dispatch = useModalPaymentDispatch(null);
+
+  // set initial number of meals to 1
+  const { fees, campaignState } = useModalPaymentState(null);
+  const [numberOfMeals, setNumberOfMeals] = useState(1);
+
+  // TODO (billy-yuan): Confirm whether mega-gam campaigns in the future will have donation matching.
+  // If so, we need to add a flag such as isMatching in the back end to determine whether to calculate a matchAmount
+
+  const calculateMatchAmount = (userDonationDollars: number) => {
+    if (
+      !campaignState &&
+      !campaignState.target_amount
+      // !isMatching
+    ) {
+      return 0;
+    }
+
+    let excess =
+      campaignState.amount_raised / 100 +
+      userDonationDollars -
+      campaignState.target_amount / 100;
+
+    return excess > 0
+      ? Math.max(0, userDonationDollars - excess)
+      : userDonationDollars;
+  };
+
+  const squareFee = fees[0];
+
+  const donationAmountDollars = numberOfMeals * props.costPerMealInDollars;
+  const matchAmount = calculateMatchAmount(donationAmountDollars);
+  const totalContribution = matchAmount + donationAmountDollars;
+  const feesAmount = squareFee
+    ? calculateFeeAmount(donationAmountDollars * 100, squareFee)
+    : 0;
+
+  const calculateMaxMeals = () => {
+    const mealFee = squareFee
+      ? calculateFeeAmount(props.costPerMealInDollars * 100, squareFee)
+      : 0;
+
+    const costPerMealPlusFee = 100 * props.costPerMealInDollars + mealFee;
+
+    return Math.floor((100 * COST_LIMIT_DOLLARS) / costPerMealPlusFee);
+  };
+
+  const updateFeesAndAmounts = (value: string) => {
+    const MAX_MEALS = calculateMaxMeals();
+    const numMeals = value ? Math.min(MAX_MEALS, parseInt(value)) : 0;
+    setNumberOfMeals(numMeals);
+
+    const donationAmountDollars = numMeals * props.costPerMealInDollars;
+
+    const newFeesAmount = squareFee
+      ? calculateFeeAmount(donationAmountDollars * 100, squareFee)
+      : 0;
+
+    const matchAmount = calculateMatchAmount(donationAmountDollars);
+
+    dispatch({
+      type: ModalPaymentConstants.SET_FEES_AMOUNT,
+      payload: newFeesAmount,
+    });
+
+    dispatch({
+      type: ModalPaymentConstants.SET_AMOUNT,
+      payload: {
+        amount: donationAmountDollars,
+        matchAmount: matchAmount,
+      },
+    });
+  };
+
+  useEffect(() => {
+    updateFeesAndAmounts(String(numberOfMeals));
+    // eslint-disable-next-line
+  }, []);
+
+  const handleMealAmount = (value: string) => {
+    updateFeesAndAmounts(value);
+  };
+
+  const openModal = (e: any) => {
+    ReactPixel.trackCustom('GiftMealPaymentNextButtonClick', {
+      numberOfMeals: numberOfMeals,
+    });
+    e.preventDefault();
+    dispatch({
+      type: ModalPaymentConstants.SET_MODAL_VIEW,
+      payload: ModalPaymentTypes.modalPages.card_details,
+    });
+  };
+
+  const subHeaderText = () => {
+    const makeSubHeaderText = (
+      amountRaisedCents: number,
```

Comment:
why are we wrapping makeSubHeaderText with subHeaderText?

Reply:
I had 2 functions that were named similarly, so I wanted to make it less confusing. I can take out `makeSubHeaderText`  and make `makeSubHeaderText` into a React component called `makeSubHeaderTextComponent` so it's clear that the purpose of each function is clearer